// expo/build/Haptic.js
import { NativeModules, Platform, Vibration } from 'react-native';
const VIBRATION_DURATION_MS = 100;
/**
 * Notification Feedback Type
 * The type of notification feedback generated by a UINotificationFeedbackGenerator object.
 * https://developer.apple.com/documentation/uikit/uinotificationfeedbacktype
 */
export var NotificationFeedbackType;
(function (NotificationFeedbackType) {
    /**
     * A notification feedback type indicating that a task has completed successfully
     */
    NotificationFeedbackType["Success"] = "success";
    /**
     * A notification feedback type indicating that a task has produced a warning
     */
    NotificationFeedbackType["Warning"] = "warning";
    /**
     * A notification feedback type indicating that a task has failed
     */
    NotificationFeedbackType["Error"] = "error";
})(NotificationFeedbackType || (NotificationFeedbackType = {}));
/**
 * Impact Feedback Style
 * The mass of the objects in the collision simulated by a UIImpactFeedbackGenerator object.
 * https://developer.apple.com/documentation/uikit/uiimpactfeedbackstyle
 */
export var ImpactFeedbackStyle;
(function (ImpactFeedbackStyle) {
    /**
     * A collision between small, light user interface elements
     */
    ImpactFeedbackStyle["Light"] = "light";
    /**
     * A collision between moderately sized user interface elements
     */
    ImpactFeedbackStyle["Medium"] = "medium";
    /**
     * A collision between large, heavy user interface elements
     */
    ImpactFeedbackStyle["Heavy"] = "heavy";
})(ImpactFeedbackStyle || (ImpactFeedbackStyle = {}));
/**
 * Triggers notification feedback.
 */
export function notification(type = NotificationFeedbackType.Success) {
    if (Platform.OS === 'ios') {
        NativeModules.ExponentHaptic.notification(type);
    }
    else {
        console.warn('`Expo.Haptic` is only available on iOS');
        Vibration.vibrate(VIBRATION_DURATION_MS, false);
    }
}
/**
 * Triggers impact feedback.
 */
export function impact(style = ImpactFeedbackStyle.Medium) {
    if (Platform.OS === 'ios') {
        NativeModules.ExponentHaptic.impact(style);
    }
    else {
        console.warn('`Expo.Haptic` is only available on iOS');
        Vibration.vibrate(VIBRATION_DURATION_MS, false);
    }
}
/**
 * Triggers selection feedback.
 */
export function selection() {
    if (Platform.OS === 'ios') {
        NativeModules.ExponentHaptic.selection();
    }
    else {
        console.warn('`Expo.Haptic` is only available on iOS');
        Vibration.vibrate(VIBRATION_DURATION_MS, false);
    }
}
